gadget/device/gpio_chardev.go

```diff
@@ -0,0 +1,88 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2025 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package device
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+	"unsafe"
+
+	"github.com/snapcore/snapd/dirs"
+	"golang.org/x/sys/unix"
+)
+
+func SnapGpioChardevPath(instanceName, plugOrSlot string) string {
+	return filepath.Join(dirs.SnapGpioChardevDir, instanceName, plugOrSlot)
+}
+
+type GPIOChardev struct {
+	Path     string
+	Name     string
+	Label    string
+	NumLines uint
+}
+
+func (c *GPIOChardev) String() string {
+	return fmt.Sprintf("(name: %s, label: %s, lines: %d)", c.Name, c.Label, c.NumLines)
+}
+
+// This has to match the memory layout of `struct gpiochip_info` found
+// in /include/uapi/linux/gpio.h in the kernel.
+type kernelChipInfo struct {
+	name, label [32]byte
+	lines       uint32
+}
+
+const _GPIO_GET_CHIPINFO_IOCTL uintptr = 0x8044b401
```

Comment:
Is the ioctl the same across all CPU architectures?

Reply:
That's a really good question, Thanks for pointing it out.

I took a look at https://elixir.bootlin.com/linux/v6.14-rc6/source/include/uapi/linux/gpio.h#L508, and it seems the only part that could vary based on architecture is the size of `struct gpiochip_info` https://elixir.bootlin.com/linux/v6.14-rc6/source/include/uapi/linux/gpio.h#L32-L36, it shouldn't be a problem as the member types specify the exact bytes don't leave room for CPU architecture to interfere.

Reply:
Wow, lucky ioctl :-)